packages/liveblocks-react/src/types/index.ts

```diff
@@ -136,142 +132,53 @@ export type CommentReactionOptions = {
   emoji: string;
 };
 
-export type ThreadsStateLoading = {
-  isLoading: true;
-  threads?: never;
-  error?: never;
-};
+export type ThreadsStateLoading = AsyncLoading<"threads">;
+export type ThreadsStateSuccess<M extends BaseMetadata> = AsyncSuccess<ThreadData<M>[], "threads">; // prettier-ignore
 
+// XXX Refactor this type away! We should NOT have threads & error state simultaneously!
+// XXX Re-express this as AsyncSuccess<ThreadData<M>[], 'threads'>
 export type ThreadsStateResolved<M extends BaseMetadata> = {
   isLoading: false;
   threads: ThreadData<M>[];
   error?: Error;
 };
 
-export type ThreadsStateSuccess<M extends BaseMetadata> = {
-  isLoading: false;
-  threads: ThreadData<M>[];
-  error?: never;
-};
-
 export type ThreadsState<M extends BaseMetadata> =
   | ThreadsStateLoading
+  // XXX Refactor this member away! We should NOT have threads & error state simultaneously!
   | ThreadsStateResolved<M>;
 
-export type InboxNotificationsStateLoading = {
-  isLoading: true;
-  inboxNotifications?: never;
-  error?: never;
-};
+export type InboxNotificationsStateLoading = AsyncLoading<"inboxNotifications">;
+export type InboxNotificationsStateError = AsyncError<"inboxNotifications">;
+export type InboxNotificationsStateSuccess = AsyncSuccess<InboxNotificationData[], "inboxNotifications">; // prettier-ignore
+export type InboxNotificationsState = AsyncResult<InboxNotificationData[], "inboxNotifications">; // prettier-ignore
 
-export type InboxNotificationsStateSuccess = {
-  isLoading: false;
-  inboxNotifications: InboxNotificationData[];
-  error?: never;
-};
+export type UnreadInboxNotificationsCountStateLoading = AsyncLoading<"count">;
+export type UnreadInboxNotificationsCountStateError = AsyncError<"count">;
+export type UnreadInboxNotificationsCountStateSuccess = AsyncSuccess<number, "count">; // prettier-ignore
+export type UnreadInboxNotificationsCountState = AsyncResult<number, "count">;
 
-export type InboxNotificationsStateError = {
-  isLoading: false;
-  inboxNotifications?: never;
-  error: Error;
-};
-
-// TODO Think about ways to remove these types as global exports
-export type InboxNotificationsState =
-  | InboxNotificationsStateLoading
-  | InboxNotificationsStateSuccess
-  | InboxNotificationsStateError;
-
-export type UnreadInboxNotificationsCountStateLoading = {
-  isLoading: true;
-  count?: never;
-  error?: never;
-};
+export type RoomNotificationSettingsStateLoading = AsyncLoading<"settings">;
+export type RoomNotificationSettingsStateError = AsyncError<"settings">;
+export type RoomNotificationSettingsStateSuccess = AsyncSuccess<RoomNotificationSettings, "settings">; // prettier-ignore
+export type RoomNotificationSettingsState = AsyncResult<RoomNotificationSettings, "settings">; // prettier-ignore
 
-export type UnreadInboxNotificationsCountStateSuccess = {
-  isLoading: false;
-  count: number;
-  error?: never;
-};
+export type HistoryVersionDataState = AsyncResult<Uint8Array>;
```

Comment:
See commit dbc6d8428edcb090bd8a70071b19dce587aa1130 that changed this. It was fine here, because the implementation would never return dual data/error values, so we could safely simplify this type here already, without implementation changes.